engine/distributed_test.go

```diff
@@ -218,6 +218,8 @@ func TestDistributedAggregations(t *testing.T) {
 		{name: "double aggregation", query: `max by (pod) (sum by (pod) (bar))`},
 		{name: "aggregation with function operand", query: `sum by (pod) (rate(bar[1m]))`},
 		{name: "binary expression with constant operand", query: `sum by (region) (bar * 60)`},
+		{name: "binary expression with distributable pairing", query: `sum by (pod) (bar * bar)`},
+		{name: "binary expression with non-distributable pairing", query: `sum by (pod) (bar * ignoring (region) bar)`},
```

Comment:
As a sanity check, could you make [this](https://github.com/thanos-io/promql-engine/pull/346/files#diff-6a4b9ef5cc84db6638cc2cf5bce5e1345aedbf74145971341be9df1ca75949c4R421) function return `true` locally and rerun the test? I want to make sure the test case fails if we distribute the binary operation.

Reply:
Added a test that fails if i mistakenly allow distribution and succeeds if i correct it again: 

```
$ go test . -run TestDistributedAggregations/.*/binary_expression_with_non-distributable_pairing/.* -failfast -test.count=1
--- FAIL: TestDistributedAggregations (0.28s)
    --- FAIL: TestDistributedAggregations/count_by___name___label#02 (0.00s)
        --- FAIL: TestDistributedAggregations/count_by___name___label#02/binary_expression_with_non-distributable_pairing (0.00s)
            --- FAIL: TestDistributedAggregations/count_by___name___label#02/binary_expression_with_non-distributable_pairing/instant/ts=121 (0.00s)
                distributed_test.go:288: Query: sum by (pod) (foo * ignoring (region, zone) xyz)
                    Explanation:
                    [*concurrencyOperator(buff=2)]:
                    └──[*aggregate] sum by ([pod]):
                       └──[*concurrencyOperator(buff=2)]:
                          └──[*dedup]:
                             └──[*coalesce]:
                                ├──[*concurrencyOperator(buff=2)]:
                                │  └──[*remoteExec] sum by (pod, zone) (foo * ignoring (region, zone) xyz) (121, 121)
                                └──[*concurrencyOperator(buff=2)]:
                                   └──[*remoteExec] sum by (pod, zone) (foo * ignoring (region, zone) xyz) (121, 121)


                distributed_test.go:289:
                distributed_test.go:290: {pod="nginx-1"} => 36 @[121000]
                testutil.go:91: distributed_test.go:291: ""

                    	exp: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{T:121000, F:36, H:(*histogram.FloatHistogram)(nil), Metric:labels.Labels{labels.Label{Name:"pod", Value:"nginx-1"}}}}, Warnings:annotations.Annotations{}}

                    	got: &promql.Result{Err:error(nil), Value:promql.Vector{}, Warnings:annotations.Annotations(nil)}

                    Diff:
                    --- Expected
                    +++ Actual
                    @@ -1,2 +1,2 @@
                    -(*promql.Result)({pod="nginx-1"} => 36 @[121000])
                    +(*promql.Result)()



FAIL
FAIL	github.com/thanos-io/promql-engine/engine	0.308s
FAIL
 fedora  ~  git  promql-engine  engine   mhoffm-distribute-binary-expressions-where-signature-contains-all-labels   
$ go test . -run TestDistributedAggregations/.*/binary_expression_with_non-distributable_pairing/.* -failfast -test.count=1
ok  	github.com/thanos-io/promql-engine/engine	0.430s
```

Since in the "count by __name__" scenario xyz and bar inhabit different remote engines; if i try to distribute it i get an empty result but if i dont it passes again.